### ISSUE TITLE
Round Not Starting (--FOUND ROOTS)

### ISSUE THREAD

**Issue by Bugleman** [XanOpiat/Python-CSGO-Cheat#78]
Ok after a bit of debugging and logging i found the error is coming from 

`y_angle = GetPlayerVars(pm, client, engine, engine_pointer, player)`

Full traceback 
Traceback (most recent call last):
  File "legit.py", line 66, in main
    player, engine_pointer, glow_manager, crosshairID, getcrosshairTarget,immunitygunganme, localTeam, crosshairTeam, y_angle = GetPlayerVars(pm, client, engine, engine_pointer, player)
  File "C:\Users\lande\Desktop\csg\Python-CSGO-Cheat-main\Classes\PlayerVars.py", line 14, in GetPlayerVars
    y_angle = pm.read_float( engine_pointer + dwClientState_ViewAngles + 0x4 )
  File "C:\Users\lande\AppData\Local\Programs\Python\Python38\lib\site-packages\pymem__init__.py", line 652, in read_float
    raise pymem.exception.MemoryReadError(address, struct.calcsize('f'), e.error_code)
pymem.exception.MemoryReadError: Could not read memory at: -58896988, length: 4 - GetLastError: 998

If you could fix that would be nice

More traceback

Traceback (most recent call last):
  File "legit.py", line 66, in main
    player, engine_pointer, glow_manager, crosshairID, getcrosshairTarget,immunitygunganme, localTeam, crosshairTeam, y_angle = GetPlayerVars(pm, client, engine, engine_pointer, player)
  File "C:\Users\lande\Desktop\csg\Python-CSGO-Cheat-main\Classes\PlayerVars.py", line 14, in GetPlayerVars
    y_angle = pm.read_float( engine_pointer + dwClientState_ViewAngles + 0x4 )
  File "C:\Users\lande\AppData\Local\Programs\Python\Python38\lib\site-packages\pymem__init.py", line 652, in read_float
    raise pymem.exception.MemoryReadError(address, struct.calcsize('f'), e.error_code)
pymem.exception.MemoryReadError: Could not read memory at: -58896988, length: 4 - GetLastError: 998
Could not read memory at: -58896988, length: 4 - GetLastError: 998
ERROR:root:Got exception on main handler
Traceback (most recent call last):
  File "C:\Users\lande\AppData\Local\Programs\Python\Python38\lib\site-packages\pymem__init.py", line 650, in read_float
    value = pymem.memory.read_float(self.process_handle, address)
  File "C:\Users\lande\AppData\Local\Programs\Python\Python38\lib\site-packages\pymem\memory.py", line 377, in read_float
    bytes = read_bytes(handle, address, struct.calcsize('f'))
  File "C:\Users\lande\AppData\Local\Programs\Python\Python38\lib\site-packages\pymem\memory.py", line 105, in read_bytes
    raise pymem.exception.WinAPIError(error_code)

**Comment by Bugleman** [comment]
tf bro

**Comment by XanOpiat** [comment]
wym tf bro? how can i fix it if i dont get any errors? how am i supposed to know whether i fixed the code or not? how can i check which variable has which value at a certain time? how can i know whether the problem has something to do with the game or with my code?

**Comment by lkean9** [comment]
Hi after a lot of checkpoints debugging, I think I know the reason. But still struggled to fix the bug. So I would like to share my findings and perhaps we can discuss about it. 

The main reason is the bit overflow issue of the variable ```engine_pointer```. I will add a few screenshots for your easier review.

![image](https://user-images.githubusercontent.com/80042476/144637399-0253cf86-723c-4a73-80b8-6c8c05c02fe4.png)

As you can see, based on the latest offests provided in the hazedumper, the ```engine``` = 1621229568, and the ```dwClientState``` = 5808076.
Then in the ```pm.read_int``` function in the line 47, the ```engine_pointer``` turns to a negative value of -66846704.

Hence, now I can confirm the problem occurs at the value of ```engine_pointer```. Then let's go to the implementation of ```pymem.read_int```

![image](https://user-images.githubusercontent.com/80042476/144638326-af2370d1-f214-491c-bfc9-eaf5daee8c9d.png)

So after a few checks, I find the main problem is how pymem deals with the byte value. From the implementation of ```pymem.memory.read_int```, I made some output and figure out the reason. The output byte value is ```b'\x10\x00\x04\xfc'```. So the negative decimal actually comes from the **little endian** of this 4-byte value: 1111 1100 0000 0100 0000 0000 0001 0000 in 2's complement is just equal to -66846704. 

![1638549234(1)](https://user-images.githubusercontent.com/80042476/144638640-192328eb-d3be-4d5e-9920-96f8be6b74a4.png)

Therefore, I think the root reason is that pymem deals with the byte using signed bit. For the memory accessing, unsigned seems more appropriate. But tbh I'm still not very familiar with the pymem. Can you think of any ways to amend this issue? Thanks!

**Comment by XanOpiat** [comment]
bro i fucking love you, thanks man i think i actually know what to do know

**Comment by XanOpiat** [comment]
what i did just know is thange the engine_pointer = pm.read_int line to pm.read_uint for me engine pointer now returns a positive value, so maybe you could test this out before i add it to the github, im still not sure what caused the cheat to work for me but that doesnt really matter does it

**Comment by lkean9** [comment]
OMFG i never believe you can reply me in just 10 fucking mins haha! Thanks a lot! I realize this issue actually because in other's C++ cheat source code, almost all the data type related to memory is unsigned: uintptr_t, uint32_t , etc, Yeah sure I will try to use uint first. I'll let you know if it works. And if it does work, then I suppose we may need to change all base/memory related function to read_uint

**Comment by XanOpiat** [comment]
yep, man i was hoping so much someone could help me figure this out thanks again

**Comment by XanOpiat** [comment]
also i dont think we have to change all the values as most of the int reads should work anyways

**Comment by lkean9** [comment]
Yes, both player and engine_pointer are negative values. I think changing for these two should be good. There will be no "round not started yet" error which means the cheat knows the round now starts. 
But then it got some error like below: i'm still trying to find the reason
![image](https://user-images.githubusercontent.com/80042476/144646751-9a74ab68-6ab7-4770-8932-7154db09edea.png)

**Comment by XanOpiat** [comment]
so this is probably in the WallhackFunctions.py file, maybe you could just change the read_ints to read_uints

**Comment by XanOpiat** [comment]
you may also try to just remove the try and except statements to check in which line the error happens

**Comment by lkean9** [comment]
OK sure. I'll try first and update you what I got! Thanks!

**Comment by XanOpiat** [comment]
nice thanks bro

**Comment by lkean9** [comment]
It finally works!!!!! OMG your code is soooooo amazing. EVERYTHING WORKS! 
All of the errors are cauased by the signed int issues. The variables that need read_uint are: engine_pointer, player, glow_manager, entity (in the for loop)

I also find that the deep root reason is actually the ```client``` and ```engine``` base memory address. Because we have different devices, we may have different addr for them? (I'm not sure though...)
In my pc, ```client = 1429340160 (0x55320000)``` and ```engine=1621229568 (0x60A20000)```, then we need to add on other offsets to get relevant memories. As long as the first hex digit increases to 8 (1xxx), the value becomes the negative. Hence uint becomes necessary. I suppose you didnt encounter this problem because your base memroy of ```client``` and ```engine``` is not big enough.

Therefore, based on my understanding, not all of users will encounter this  damned bug, but to make it more robust, change all read_int to read_uint may solve the problem for good

**Comment by lkean9** [comment]
I can also pull a request for your easy merge if you want. It;s just I added shit ton of checkpoints lol so need some time to restore the code. anyway if needed just le me know. Thanks for your great code! Actually I also used to make a even bad ass code for third-party server like 5E, b5 or even faceit. Ofc I know it would be much more difficult, the cost may be much higher (money & time) so if you are willing, perhaps we can also discuss some stuff in the future lol
